File: src/_u_setops_c.c
/* **********************************************************

   Code is used in gRbase
   
   Functions for set operations, most notably maxset and minset
   
   * maxset: Finds the maximal elements of a set, i.e. elements
   contained in other elements are discarded are marked with a 0,
   elements not contained in other sets are marked with a 1.
   
   * minset: Does the opposite of maxset
   
   * maxset and minset are used in removeRedundant in R
   
   * isin checks if an element (a vector) is contained in a set 
   (a list)
   
   * the subsetof-functions are internal.
   
   Created while in Bristol, 2007/2008
   Søren Højsgaard

** **********************************************************/

#include <string.h>
#include <Rdefines.h>

void C_subsetof1(char **setxsety, int *sx, int *ex, int *sy, int *ey, int *ans);
void C_subsetof2(char **qset, int *len_qset, 
		 char **setlist, int *start_set, int *end_set, int *ans);
void C_maxset(char **setlist, int *ends, int *nset, int *keepvec);
void C_minset(char **setlist, int *ends, int *nset, int *keepvec);

void C_isin(char **qset, int *len_qset, char **setlist, int *ends, int *len_setlist, int *keepvec);

/*
  Returns vector with entries 0 meaning that the elements are
  contained in other elements while 1 means that they are maximal
*/
void C_maxset(char **setlist, int *ends, int *nset, int *keepvec)
{
  
  int ii, jj;
  int ans;

  int starts[*nset];
  starts[0]=1;
  keepvec[0]=1;
  for (ii=1; ii<*nset; ii++){ 
    starts[ii] = ends[ii-1]+1;
    keepvec[ii] = 1;
  }

  for (ii=0; ii<*nset-1; ii++){ 
    if (keepvec[ii]==1){
      /* Rprintf("SET 1: %i %i %i\n", ii, starts[ii], ends[ii]);    */
      for (jj=ii+1; jj<*nset; jj++){
	if (keepvec[jj]==1){
	  /* Rprintf("  SET 2: %i %i %i\n", jj, starts[jj], ends[jj]);   */
	  C_subsetof1(setlist, &starts[jj], &ends[jj], &starts[ii], &ends[ii], &ans); 
	  /* Rprintf("  ans: %i \n", ans);    */
	  if (ans==1)
	    keepvec[jj] = 0;
	  else
	    keepvec[jj] = 1;
	}
      }
    }
  }
}


void C_minset(char **setlist, int *ends, int *nset, int *keepvec)
{
  
  int ii, jj;
  int ans;

  int starts[*nset];
  starts[0]=1;
  keepvec[0]=1;
  for (ii=1; ii<*nset; ii++){ 
    starts[ii] = ends[ii-1]+1;
    keepvec[ii] = 1;
  }

  for (ii=0; ii<*nset-1; ii++){ 
    if (keepvec[ii]==1){
      /*       Rprintf("SET 1: no: %i start: %i end: %i\n", ii, starts[ii], ends[ii]);     */
      for (jj=ii+1; jj<*nset; jj++){
	if (keepvec[jj]==1){
	  /* 	  Rprintf("  SET 2: %i %i %i\n", jj, starts[jj], ends[jj]);    */
	  C_subsetof1(setlist, &starts[ii], &ends[ii], &starts[jj], &ends[jj],  &ans); 
	  /* 	  Rprintf("  ans: %i \n", ans);     */
	  if (ans==1)
	    keepvec[jj] = 0;
	  else
	    keepvec[jj] = 1;
	}
      }
    }
  }
}



/* ************************************************************ */
/* ************************************************************ */


/* C_subsetof1: 
   Returns 0 if x is subset of y and 1 otherwise.
   - setxsety is c(setx, sety)
   - x and y are defined through the indices (sx,ex) and (sy,ey) in setxsety, 
   for example 1,3,4,7 says setx is setxsety[1:3] and sety is setxsety[4:7]
   - Notice that indexing follows R-standard, i.e. starts from 1      */
void C_subsetof1(char **setxsety, int *sx, int *ex, int *sy, int *ey, int *ans)
{
/*   Rprintf("    sx %i ex %i sy %i ey %i\n", *sx, *ex, *sy, *ey ); */
  int ii, jj, nmatch=0, nx;
  
  nx   = *ex-*sx+1;
  *ans = 0;
  
  for (ii=*sx-1; ii<*ex; ii++){
    /* Rprintf("    ii %i %s\n", ii, setxsety[ii]);  */
    for (jj=*sy-1; jj<*ey; jj++){
      /* Rprintf("      jj %i %s\n", jj, setxsety[jj]);  */
      if (!strcmp(setxsety[ii],setxsety[jj])){
	nmatch++;
	/* 	Rprintf("  match...%i %i %s %s %i\n", ii, jj, setxsety[ii], setxsety[jj], nmatch);   */
	break;
      }      
    }
    if (nmatch == nx){
      *ans = 1;
      break;
    }
  }
/*    Rprintf(" ans:  %i \n", *ans);   */
}




/* C_subsetof2:
   Returns 1 if qset (queryset) is contained in the subset of setlist defined by
   (start_set, end_set).
*/
void C_subsetof2(char **qset, int *len_qset, 
		 char **setlist, int *start_set, int *end_set, int *ans)
{
  /*   Rprintf("    start_set %i end_set %i sy %i ey %i\n", *start_set, *end_set, *sy, *ey ); */
  int ii, jj, nmatch=0; //, nx;
  
  //nx   = *end_set-*start_set+1;
  *ans = 0;
  
  for (ii=0; ii<*len_qset; ii++){
    /* Rprintf("    ii %i %s\n", ii, setlist[ii]);   */
    for (jj=*start_set-1; jj<*end_set; jj++){
      /* Rprintf("      jj %i %s\n", jj, setlist[jj]);   */
      if (!strcmp(qset[ii],setlist[jj])){
	nmatch++;
	/* Rprintf("  match...%i %i %s %s %i\n", ii, jj, setlist[ii], setlist[jj], nmatch);   */
	break;
      }      
    }
    if (nmatch == *len_qset){
      *ans = 1;
      break;
    }
  }
  /* Rprintf(" ans:  %i \n", *ans);    */
}

/* ************************************************************ */
/* ************************************************************ */


void C_isin(char **qset, int *len_qset, char **setlist, int *ends, int *len_setlist, int *keepvec)
{
  
  int ii, ans, starts[*len_setlist];

  starts[0]=1;
  keepvec[0]=1;
  for (ii=1; ii<*len_setlist; ii++){ 
    starts[ii]  = ends[ii-1]+1;
    keepvec[ii] = 1;
  }

  for (ii=0; ii<*len_setlist; ii++){ 
    C_subsetof2(qset, len_qset, setlist, &starts[ii], &ends[ii], &ans); 
    keepvec[ii] = ans;
  }
}

File: R/u_setops-c.R
## #############################################################
##
#' @title Suite of set operations
#' @description Set operations for gRbase and related packages.
#' @name set_operations
#' @author Søren Højsgaard, \email{sorenh@@math.aau.dk}
##
## #############################################################
#' 
#' @param x,set,set2 Vector representing a set.
#' @param setlist List of vectors (representing a set of subsets)
#' @param maximal Logical; see section 'Details' for a description.
#' @param all Logical; see section 'Details' for a description.
#' @param index Logical; should indices (in setlist) be returned or a
#'     set of subsets.
#'
#' @details
#'
#'  'setlist' is a list of vectors representing a set of subsets;
#'  i.e. V1,...VQ where Vk is a subset of some base set V.
#'
#'  'all' If true, \code{get_superset} will return index of all
#'  vectors containing the element; otherwise only the first index is
#'  returned.
#'
#'  \code{is_inset}: Checks if the set
#'  x is in one of the Vk's.
#'
#'  \code{remove_redundant}: Returns those Vk which are not contained
#'  in other subsets; i.e. gives the maximal sets. If maximal is FALSE
#'  then returns the minimal sets; i.e. Vk is returned if Vk is
#'  contained in one of the other sets Vl and there are no set Vn
#'  contained in Vk.
#'  
#'  Notice that the comparisons are made by turning the elements into
#'  characters and then comparing these. Hence 1 is identical to "1".
#'
#' 
#' @examples
#'
#' set <- list(c(1, 2), c(1, 2, 3), c(2, 3, 6), c(2, 4), c(5, 6), 5)            
#'                                                             
#' el1 <- c(2, 1)                                               
#' el2 <- c(2, 3)                                               
#' el3 <- c(4, 3)                                               
#' el4 <- c(2, 1, 3)                                             
#'                                                             
#' maximal_sets(set)                                           
#' minimal_sets(set)                                           
#'                                                             
#' remove_redundant(set)                                       
#' remove_redundant(set, maximal=FALSE)                        
#'                                                             
#' is_inset(el1, set)                                          
#' is_inset(el2, set)                                          
#' is_inset(el3, set)                                          
#'                                                             
#' get_subset(el1, set)
#' get_subset(el1, set)                                        
#' get_subset(el2, set)                                        
#' get_subset(el3, set)                                        
#'
#' get_superset(el1, set)                                      
#' get_superset(el1, set, all=TRUE)                                      
#' get_superset(el2, set)                                      
#' get_superset(el3, set)                                      
## #'                                                             
## #' is_subsetof(el1, el1)                                       
## #' is_subsetof(el1, el2)                                       
## #' is_subsetof(el1, el4)
#' 


#' @export
#' @rdname set_operations
maximal_sets <- function(setlist, index=FALSE){
    if (length(setlist)<=1){
        if (index) return(1)
        else return(setlist)
    }
    
    lenx     <- c(lapply(setlist,length), recursive=TRUE)
    ooo      <- order(lenx, decreasing=TRUE)
    setlist2 <- setlist[ooo]
    ends     <- cumsum(c( lapply(setlist2,length), recursive=TRUE ))
    iii<-.C("C_maxset",
            setlist=as.character(c(setlist2, recursive=TRUE)),
            ends=ends, nset=length(setlist2), ans=integer(length(setlist2))
          , PACKAGE="gRbase")$ans
    iii <- iii[order(ooo)]
    
    if (index){
        iii
    } else {
        setlist[iii==1]
    }
}

#' @export
#' @rdname set_operations
minimal_sets <- function(setlist, index=FALSE){
    if (length(setlist) <= 1){
        if (index) return(1)
        else return(setlist)
    }
    
    lenx     <- c(lapply(setlist,length), recursive=TRUE)
    ooo      <- order(lenx, decreasing=FALSE)
    setlist2 <- setlist[ooo]
    ends     <- cumsum(c( lapply(setlist2,length), recursive=TRUE ))
    iii<-.C("C_minset",
            setlist=as.character(c(setlist2, recursive=TRUE)),
            ends=ends, nset=length(setlist2), ans=integer(length(setlist2))
          , PACKAGE="gRbase")$ans
    iii <- iii[order(ooo)]
    
    if (index){
        iii
    } else {
        setlist[iii==1]
    }
}

## A function to remove redundant generators.  If maximal=T, returns
## the maximal generators, if =F, the minimal generators.
## Can be speeded up if the as.character part can be avoided...

#' @export
#' @rdname set_operations
remove_redundant <- function(setlist, maximal=TRUE, index=FALSE){
  if (maximal) maximal_sets(setlist, index)
  else minimal_sets(setlist, index)
}

## Is x contained in any vector in setlist;
#' @export
#' @rdname set_operations
is_inset <- function(x, setlist, index=FALSE){
  .isin(setlist, x, index)
}

.isin <- function(setlist, x, index=FALSE){
    len.setlist <- length(setlist)
    if (len.setlist == 0){
        if (index) return(0)
        else return(FALSE)
    }

    ## FIXME: Looks strange
    if (len.setlist < 1){
        if (index) return(rep(1, length(x)))
        else return(TRUE)
    }
    
    ll    <- cumsum(c(lapply(setlist, length), recursive=TRUE))
    iii<-.C("C_isin",
            as.character(x), length(x),
            as.character(c(setlist, recursive=TRUE)), ll, len.setlist,
            ans=integer(len.setlist)
          , PACKAGE="gRbase")$ans
    
    if (index) { return(iii) }
    else { return(any(iii)) }
}

#' @export
#' @rdname set_operations
get_subset <- get_subset_

#' @export
#' @rdname set_operations
get_superset <- get_superset_

#' @export
#' @rdname set_operations
is_subsetof <- is_subsetof_


## FIXME: is.subsetof : Use Rcpp implementation

#' @export
#' @rdname set_operations
is.subsetof <- function(x, set){
  all(match(x, set) > 0)
}

## ###################################################################
##
#' @title Create all subsets
#' @description Create all subsets of a vector
#' @name all_subsets
#' @author Søren Højsgaard, \email{sorenh@@math.aau.dk}
##
## Issues: deprecate allSubsets; use all_subsets instead
##
## ###################################################################
#' @param x Vector

#' @export
#' @rdname all-subsets
all_subsets <- allSubsets_

#' @export
#' @rdname all-subsets
all_subsets0 <- allSubsets0_


.all_subsets0 <- function(x) {
    y <- list(vector(mode(x), length = 0))
    for (i in seq_along(x)) {
        y <- c(y, lapply(y, "c", x[i]))
    }
    y[-1L]
}

.all_subsets <- function(x){
    out <- vector("list", length=2^length(x))
    ny = 1 # filled elements of out
    for (i in seq_along(x)){
        z = x[i]
        for (k in 1:ny){
            out[[ny + k]] = c(out[[k]],z)
        }
        ny = 2 * ny
    }
    out[-1]
}



## #' @rdname all-subsets
## #' @param g.sep Pick a value which is not in x
## allSubsets <- function(x, g.sep="+"){
##   if (length(x)==1)
##     return(x)
##   else {
##     val <- x[1]
##     for (i in 2:length(x)){
##       v <- paste(val,x[i],sep=g.sep)
##       val <- c(val,x[i],v)
##     }
##     val <- strsplit(val,paste("\\",g.sep,sep=""))
##     return(val)
##   }
## }





## ###################################################################
##
#' @title Create all possible pairs
#' @description Create all possible pairs of two character vectors.
#' @name all_pairs
#' @author Søren Højsgaard, \email{sorenh@@math.aau.dk}
#'
## ###################################################################
#'
#' @param x,y Character vectors.
#' @param sort Logical.
#' @param result A list or a matrix.
#'
#' @details NOTICE: If y is not NULL then x and y must be disjoint (no
#'     checks are made); otherwise pairs of identical elements wil also be obtained. 
#'
#' @examples
#'
#' x <- letters[1:4]
#' y <- letters[5:7]
#'
#' all_pairs(x)
#' all_pairs(x, result="matrix")
#'
#' all_pairs(x, y)
#' all_pairs(x, y, result="matrix")

#' @export
#' @rdname all_pairs
all_pairs <- all_pairs__

## FIXME names2pairs should be deprecated and replaced by all_pairs
#' @export
#' @rdname all_pairs
names2pairs <- function(x, y=NULL, sort=TRUE, result="list"){
  result <- match.arg(result, c("list", "matrix"))
  lenx <- length(x)
  leny <- length(y)

  if (leny == 0){
    if (lenx == 1){
      if (result == "matrix")
        return(matrix(nrow=0, ncol=2))
      else
        return(list())
    } else {
      cc   <- combn_prim(1:length(x), 2)
      out  <- x[cc]
      dim(out) <- dim(cc)
      if (sort){
        idx <- out[1,] > out[2, ]
        out[1:2,idx] <- out[2:1, idx]
      }
      if (result == "matrix")
        return(t.default(out))
      else
        return(colmat2list(out))
    }
  } else {
    out <- cbind(rep(x, each=leny), rep(y, times=lenx))
    if (sort){
      idx <- out[,1] > out[,2]
      out[idx, 1:2] <- out[idx, 2:1]
    }

    if (identical(result, "matrix")) out 
    else rowmat2list__(out)
  }
}



